File: ajedrez.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;

public class Ajedrez extends JPanel implements MouseListener {

    private static final int TAM = 60;
    private static final String BLANCO = "white";
    private static final String NEGRO = "black";

    private Pieza[][] tablero = new Pieza[8][8];
    private String turno = BLANCO;

    private int selF = -1, selC = -1;
    private java.util.List<Point> movimientos = new java.util.ArrayList<>();

    public Ajedrez() {
        inicializarTablero();
        addMouseListener(this);
        setPreferredSize(new Dimension(480, 480));
    }

    private void inicializarTablero() {
        String[] orden = {"torre", "caballo", "alfil", "reina", "rey", "alfil", "caballo", "torre"};

        for (int i = 0; i < 8; i++) {
            tablero[0][i] = new Pieza(orden[i], NEGRO);
            tablero[1][i] = new Pieza("peon", NEGRO);
            tablero[6][i] = new Pieza("peon", BLANCO);
            tablero[7][i] = new Pieza(orden[i], BLANCO);
        }
    }

    @Override
    public void paintComponent(Graphics g) {
        super.paintComponent(g);

        // Dibujar casillas
        for (int f = 0; f < 8; f++) {
            for (int c = 0; c < 8; c++) {
                boolean claro = (f + c) % 2 == 0;
                g.setColor(claro ? new Color(238, 238, 210) : new Color(118, 150, 86));
                g.fillRect(c * TAM, f * TAM, TAM, TAM);
            }
        }

        // Movimientos posibles
        g.setColor(Color.BLUE);
        for (Point p : movimientos) {
            g.drawRect(p.y * TAM, p.x * TAM, TAM, TAM);
            g.drawRect(p.y * TAM + 1, p.x * TAM + 1, TAM - 2, TAM - 2);
        }

        // Dibujar piezas
        g.setFont(new Font("Arial", Font.PLAIN, 40));
        for (int f = 0; f < 8; f++) {
            for (int c = 0; c < 8; c++) {
                Pieza p = tablero[f][c];
                if (p != null) {
                    g.drawString(simbolo(p), c * TAM + 15, f * TAM + 45);
                }
            }
        }
    }

    private String simbolo(Pieza p) {
        return switch (p.tipo + p.color) {
            case "reywhite" -> "♔";
            case "reinawhite" -> "♕";
            case "torrewhite" -> "♖";
            case "alfilwhite" -> "♗";
            case "caballowhite" -> "♘";
            case "peonwhite" -> "♙";

            case "reyblack" -> "♚";
            case "reinablack" -> "♛";
            case "torreblack" -> "♜";
            case "alfilblack" -> "♝";
            case "caballoblack" -> "♞";
            case "peonblack" -> "♟";

            default -> "?";
        };
    }

    // ------------------ MOVIMIENTO ------------------

    @Override
    public void mouseClicked(MouseEvent e) {
        int f = e.getY() / TAM;
        int c = e.getX() / TAM;

        if (selF != -1) {
            // Intentar mover
            if (movimientos.contains(new Point(f, c))) {
                mover(selF, selC, f, c);
                turno = turno.equals(BLANCO) ? NEGRO : BLANCO;
            }
            selF = selC = -1;
            movimientos.clear();
        } else {
            Pieza p = tablero[f][c];
            if (p != null && p.color.equals(turno)) {
                selF = f;
                selC = c;
                movimientos = obtenerMovimientosValidos(f, c);
            }
        }

        repaint();
    }

    private void mover(int f0, int c0, int f1, int c1) {
        Pieza pieza = tablero[f0][c0];
        Pieza destino = tablero[f1][c1];

        if (destino != null && destino.tipo.equals("rey")) {
            JOptionPane.showMessageDialog(this, "¡Has ganado!\nPiezas " + pieza.color.toUpperCase());
            System.exit(0);
        }

        tablero[f1][c1] = pieza;
        tablero[f0][c0] = null;

        // Promoción
        if (pieza.tipo.equals("peon") && (f1 == 0 || f1 == 7)) {
            promocionar(f1, c1, pieza.color);
        }
    }

    private void promocionar(int f, int c, String color) {
        String[] opciones = {"reina", "torre", "alfil", "caballo"};
        String tipo = (String) JOptionPane.showInputDialog(
                this, "Promociona tu peón", "Promoción",
                JOptionPane.PLAIN_MESSAGE, null, opciones, opciones[0]
        );
        tablero[f][c] = new Pieza(tipo, color);
    }

    // ------------------ MOVIMIENTOS ------------------

    private java.util.List<Point> obtenerMovimientosValidos(int f, int c) {
        java.util.List<Point> lista = new java.util.ArrayList<>();

        for (int f1 = 0; f1 < 8; f1++) {
            for (int c1 = 0; c1 < 8; c1++) {
                if (movimientoValido(f, c, f1, c1)) {
                    lista.add(new Point(f1, c1));
                }
            }
        }
        return lista;
    }

    private boolean movimientoValido(int f0, int c0, int f1, int c1) {
        Pieza p = tablero[f0][c0];
        Pieza destino = tablero[f1][c1];

        if (destino != null && destino.color.equals(p.color))
            return false;

        int df = f1 - f0;
        int dc = c1 - c0;

        switch (p.tipo) {

            case "peon" -> {
                int d = p.color.equals(BLANCO) ? -1 : 1;
                int filaInicio = p.color.equals(BLANCO) ? 6 : 1;

                if (dc == 0 && df == d && destino == null)
                    return true;

                if (f0 == filaInicio && dc == 0 && df == 2 * d &&
                        destino == null && tablero[f0 + d][c0] == null)
                    return true;

                if (Math.abs(dc) == 1 && df == d && destino != null)
                    return true;
            }

            case "torre" -> {
                if (df == 0 || dc == 0)
                    return caminoLibre(f0, c0, f1, c1);
            }

            case "alfil" -> {
                if (Math.abs(df) == Math.abs(dc))
                    return caminoLibre(f0, c0, f1, c1);
            }

            case "reina" -> {
                if (df == 0 || dc == 0 || Math.abs(df) == Math.abs(dc))
                    return caminoLibre(f0, c0, f1, c1);
            }

            case "caballo" -> {
                return (Math.abs(df) == 1 && Math.abs(dc) == 2) ||
                       (Math.abs(df) == 2 && Math.abs(dc) == 1);
            }

            case "rey" -> {
                return Math.max(Math.abs(df), Math.abs(dc)) == 1;
            }
        }

        return false;
    }

    private boolean caminoLibre(int f0, int c0, int f1, int c1) {
        int df = Integer.compare(f1, f0);
        int dc = Integer.compare(c1, c0);

        int pasos = Math.max(Math.abs(f1 - f0), Math.abs(c1 - c0));

        for (int i = 1; i < pasos; i++) {
            if (tablero[f0 + df * i][c0 + dc * i] != null)
                return false;
        }
        return true;
    }

    // ------------------ MAIN ------------------

    public static void main(String[] args) {
        JFrame ventana = new JFrame("Ajedrez - 2 Jugadores");
        ventana.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        ventana.add(new Ajedrez());
        ventana.pack();
        ventana.setVisible(true);
    }

    // Métodos vacíos del MouseListener
    public void mousePressed(MouseEvent e) {}
    public void mouseReleased(MouseEvent e) {}
    public void mouseEntered(MouseEvent e) {}
    public void mouseExited(MouseEvent e) {}
}

File: pruebas/ajedrez.py
import tkinter as tk
from tkinter import messagebox

TAM = 60
BLANCO = "white"
NEGRO = "black"


class Pieza:
    def __init__(self, tipo, color):
        self.tipo = tipo
        self.color = color


class Ajedrez:
    def __init__(self, root):
        self.root = root
        self.root.title("Ajedrez - 2 Jugadores")

        self.turno = BLANCO
        self.seleccion = None
        self.movimientos_posibles = []
        self.tablero = [[None for _ in range(8)] for _ in range(8)]

        self.canvas = tk.Canvas(root, width=640, height=640)
        self.canvas.pack()

        self.inicializar_tablero()
        self.dibujar_tablero()

        self.canvas.bind("<Button-1>", self.click)

    def inicializar_tablero(self):
        orden = ["torre", "caballo", "alfil", "reina", "rey", "alfil", "caballo", "torre"]
        for i in range(8):
            self.tablero[0][i] = Pieza(orden[i], NEGRO)
            self.tablero[1][i] = Pieza("peon", NEGRO)
            self.tablero[6][i] = Pieza("peon", BLANCO)
            self.tablero[7][i] = Pieza(orden[i], BLANCO)

    def simbolo(self, pieza):
        simbolos = {
            ("rey", BLANCO): "♔", ("reina", BLANCO): "♕",
            ("torre", BLANCO): "♖", ("alfil", BLANCO): "♗",
            ("caballo", BLANCO): "♘", ("peon", BLANCO): "♙",
            ("rey", NEGRO): "♚", ("reina", NEGRO): "♛",
            ("torre", NEGRO): "♜", ("alfil", NEGRO): "♝",
            ("caballo", NEGRO): "♞", ("peon", NEGRO): "♟",
        }
        return simbolos[(pieza.tipo, pieza.color)]

    def dibujar_tablero(self):
        self.canvas.delete("all")

        for f in range(8):
            for c in range(8):
                color = "#EEEED2" if (f + c) % 2 == 0 else "#769656"
                self.canvas.create_rectangle(
                    c * TAM, f * TAM, (c + 1) * TAM, (f + 1) * TAM, fill=color
                )

        for (f, c) in self.movimientos_posibles:
            self.canvas.create_rectangle(
                c * TAM, f * TAM, (c + 1) * TAM, (f + 1) * TAM,
                outline="blue", width=4
            )

        for f in range(8):
            for c in range(8):
                pieza = self.tablero[f][c]
                if pieza:
                    self.canvas.create_text(
                        c * TAM + TAM // 2,
                        f * TAM + TAM // 2,
                        text=self.simbolo(pieza),
                        font=("Arial", 32)
                    )

    def click(self, event):
        f, c = event.y // TAM, event.x // TAM

        if self.seleccion:
            if (f, c) in self.movimientos_posibles:
                self.mover(*self.seleccion, f, c)
                self.turno = NEGRO if self.turno == BLANCO else BLANCO
            self.seleccion = None
            self.movimientos_posibles = []
        else:
            pieza = self.tablero[f][c]
            if pieza and pieza.color == self.turno:
                self.seleccion = (f, c)
                self.movimientos_posibles = self.obtener_movimientos_validos(f, c)

        self.dibujar_tablero()

    def mover(self, f0, c0, f1, c1):
        pieza = self.tablero[f0][c0]
        destino = self.tablero[f1][c1]

        if destino and destino.tipo == "rey":
            messagebox.showinfo(
                "Fin del juego",
                f"¡Has ganado!\nPiezas {pieza.color.upper()}"
            )
            self.root.quit()

        self.tablero[f1][c1] = pieza
        self.tablero[f0][c0] = None

        if pieza.tipo == "peon" and (f1 == 0 or f1 == 7):
            self.promocionar_peon(f1, c1, pieza.color)

    def promocionar_peon(self, f, c, color):
        opciones = ["reina", "torre", "alfil", "caballo"]
        ventana = tk.Toplevel(self.root)
        ventana.title("Promoción de peón")

        def elegir(tipo):
            self.tablero[f][c] = Pieza(tipo, color)
            ventana.destroy()
            self.dibujar_tablero()

        for o in opciones:
            tk.Button(
                ventana,
                text=o.capitalize(),
                width=10,
                command=lambda t=o: elegir(t)
            ).pack(padx=10, pady=5)

    def obtener_movimientos_validos(self, f, c):
        return [(f1, c1) for f1 in range(8) for c1 in range(8)
                if self.movimiento_valido(f, c, f1, c1)]

    def movimiento_valido(self, f0, c0, f1, c1):
        pieza = self.tablero[f0][c0]
        destino = self.tablero[f1][c1]

        if destino and destino.color == pieza.color:
            return False

        df, dc = f1 - f0, c1 - c0

        if pieza.tipo == "peon":
            d = -1 if pieza.color == BLANCO else 1
            fila_inicio = 6 if pieza.color == BLANCO else 1

            # Movimiento simple
            if dc == 0 and df == d and destino is None:
                return True

            # Movimiento doble (NUEVO)
            if (f0 == fila_inicio and dc == 0 and df == 2 * d and
                    destino is None and self.tablero[f0 + d][c0] is None):
                return True

            # Captura
            if abs(dc) == 1 and df == d and destino:
                return True

        if pieza.tipo == "torre":
            return (df == 0 or dc == 0) and self.camino_libre(f0, c0, f1, c1)

        if pieza.tipo == "alfil":
            return abs(df) == abs(dc) and self.camino_libre(f0, c0, f1, c1)

        if pieza.tipo == "reina":
            return (df == 0 or dc == 0 or abs(df) == abs(dc)) and self.camino_libre(f0, c0, f1, c1)

        if pieza.tipo == "caballo":
            return (abs(df), abs(dc)) in [(1, 2), (2, 1)]

        if pieza.tipo == "rey":
            return max(abs(df), abs(dc)) == 1

        return False

    def camino_libre(self, f0, c0, f1, c1):
        df = f1 - f0
        dc = c1 - c0
        pasos = max(abs(df), abs(dc))
        df //= pasos if pasos else 1
        dc //= pasos if pasos else 1

        for i in range(1, pasos):
            if self.tablero[f0 + df * i][c0 + dc * i]:
                return False
        return True


if __name__ == "__main__":
    root = tk.Tk()
    Ajedrez(root)
    root.mainloop()
